From d6d4ae7c6a755fb7e7b65726a544b48c50b04f2f Mon Sep 17 00:00:00 2001
From: remittor <remittor@gmail.com>
Date: Thu, 25 Feb 2016 10:31:05 +0200
Subject: [PATCH] Add sleep commands to update script

---
 tools/releasetools/edify_generator.py | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/tools/releasetools/edify_generator.py b/tools/releasetools/edify_generator.py
index a6baf64..ea43f81 100644
--- a/tools/releasetools/edify_generator.py
+++ b/tools/releasetools/edify_generator.py
@@ -189,6 +189,7 @@ class EdifyGenerator(object):
       self.script.append('mount("%s", "%s", "%s", "%s");' %
                          (p.fs_type, common.PARTITION_TYPES[p.fs_type],
                           p.device, p.mount_point))
+      self.script.append('sleep(2);')
       self.mounts.add(p.mount_point)
 
   def Unmount(self, mount_point):
@@ -196,11 +197,13 @@ class EdifyGenerator(object):
     if mount_point in self.mounts:
       self.mounts.remove(mount_point)
       self.script.append('unmount("%s");' % (mount_point,))
+      self.script.append('sleep(2);')
 
   def UnpackPackageDir(self, src, dst):
     """Unpack a given directory from the OTA package into the given
     destination directory."""
     self.script.append('package_extract_dir("%s", "%s");' % (src, dst))
+    self.script.append('sleep(12);')
 
   def Comment(self, comment):
     """Write a comment into the update script."""
@@ -224,6 +227,7 @@ class EdifyGenerator(object):
       self.script.append('format("%s", "%s", "%s", "%s", "%s");' %
                          (p.fs_type, common.PARTITION_TYPES[p.fs_type],
                           p.device, p.length, p.mount_point))
+      self.script.append('sleep(2);')
 
   def DeleteFiles(self, file_list):
     """Delete all files in file_list."""
@@ -331,6 +335,7 @@ class EdifyGenerator(object):
     for p in sorted(self.mounts):
       self.script.append('unmount("%s");' % (p,))
     self.mounts = set()
+    self.script.append('sleep(6);')
 
   def AddToZip(self, input_zip, output_zip, input_path=None):
     """Write the accumulated script to the output_zip file.  input_zip
-- 
1.9.5.msysgit.0

